File: IT/2/Multithreading/Esercizi/Java/Esercizio26.java
/* file Esercizio26.java */

import java.util.Deque;
import java.util.LinkedList;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import java.util.concurrent.Semaphore;
 
public class Esercizio26 {
	private final int BUFFER_SIZE = 25;
	private Lock lock = new ReentrantLock();
	private Semaphore full = new Semaphore(0);
	private Semaphore empty = new Semaphore(BUFFER_SIZE);
	private Deque<String> buffer = new LinkedList<>();

	// thread produttore
	class Producer implements Runnable {
		@Override
		public void run() {
			Thread t = Thread.currentThread();
			while (true) {
				try {
					String item = "pacco regalo";
					empty.acquire();
					lock.lock();
					// aggiunge un nuovo pacco regalo sullo scaffale
					buffer.addFirst(item);
					lock.unlock();
					System.out.println(t.getName() + " prepara un pacco regalo (" + buffer.size() + " disponibili).");
					full.release();

					int pause = ThreadLocalRandom.current().nextInt(1, 25+1);
					t.sleep(pause*1000);
				}
				catch (InterruptedException e) {
					e.printStackTrace();
				}
			}
		}
	}

	// thread consumatore
	class Consumer implements Runnable {
		@Override
		public void run() {
			Thread t = Thread.currentThread(); 
			while (true) {
				try {
					full.acquire();

					// preleva un pacco regalo dallo scaffale
					lock.lock();
					String item = buffer.removeFirst();
					lock.unlock();

					System.out.println(t.getName() + " acquista un " + item + " (" + buffer.size() + " disponibili).");
					empty.release();

					int pause = ThreadLocalRandom.current().nextInt(1, 25+1);
					t.sleep(pause*1000);
				}
				catch (InterruptedException e) {
					e.printStackTrace();
				}
			}
		}
	}

	/* il costruttore avvia l'applicazione */

	public Esercizio26() {
		new Thread(new Producer()).start();
		new Thread(new Consumer()).start();
	}

	public static void main(String [] args) {
		new Esercizio26();
	}
}

File: IT/2/Multithreading/Esercizi/Python/esercizio26.py
# file esercizio26.py

from __future__ import print_function
from multiprocessing import Process, current_process, Semaphore, Lock, Manager
from random import randint
from time import sleep

def producer(fill, empty, buffLock, buff):
	p = current_process().name

	while True:
		item = "pacco regalo"

		empty.acquire()
		buffLock.acquire()
		# aggiunge un nuovo pacco regalo sullo scaffale
		buff.put(item)
		buffLock.release()

		print("{} prepara un pacco regalo ({} disponibili).".format(p, buff.qsize()))
		fill.release()

		pause = randint(1, 25)
		sleep(pause)

def consumer(fill, empty, buffLock, buff):
	p = current_process().name
	while True:
		fill.acquire()
		buffLock.acquire()
		# preleva un pacco regalo dallo scaffale
		item = buff.get()
		buffLock.release()

		print("{} acquista un {} ({} rimanenti).".format(p, item, buff.qsize()))
		empty.release()

		pause = randint(1, 25)
		sleep(pause)

if __name__ == "__main__":
	buff = Manager().Queue(maxsize=5)
	buffLock = Lock()
	fill = Semaphore(0)
	empty = Semaphore(25)

	pProducer = Process(target=producer, args=(fill, empty, buffLock, buff,))
	pProducer.start()

	pConsumer = Process(target=consumer, args=(fill, empty, buffLock, buff,))
	pConsumer.start()

	pProducer.join()
	pConsumer.join()

